Implement a common framework for writing CVN modules for different larsoft geometries

nitish-nayak opened the pull request:
Duplicate of [larrecodnn#34](https://github.com/LArSoft/larrecodnn/pull/34) but now a PR from a feature branch. 

Aims to integrate the DUNE CVN pre-processing infrastructure into larrecodnn. In the process, also rationalizes some of the dune code.

- Common framework for producing pixel maps by deriving from a base class, rather than rewriting and adding
- Better structures for saving auxillary information for training

More details are given here : https://indico.fnal.gov/event/56692/contributions/252666/attachments/160682/211783/10_18_22_larcvn.pdf

FNALbuild commented:
A new Pull Request was created by  @nitish-nayak (Nitish Nayak) for develop.

It involves the following packages:

larrecodnn


@LArSoft/level-1-managers, @LArSoft/level-2-managers can you please review it and eventually sign? Thanks.

cms-bot commands are listed <a href="https://cdcvs.fnal.gov/redmine/projects/larsoft/wiki/Pull_request_comments_that_trigger_CI_actions">here</a>

FNALbuild commented:
The code-checks are being triggered in jenkins.

FNALbuild commented:
-code-checks
Pull request failed code-formatting checks. Please ensure that `cetmodules` has been setup and execute the following command from the top-level directory of your repository:

```
format-code \
  larrecodnn/CVN/func/CVNImageUtils.cxx \ 
  larrecodnn/CVN/func/LArTrainingData.cxx \ 
  larrecodnn/CVN/func/classes.h \ 
  larrecodnn/CVN/modules/LArNuCVNZlibMaker_module.cc
```
Then commit the changes and push them to your PR branch.

FNALbuild commented:
Pull request #35 was updated. @LArSoft/level-1-managers, @LArSoft/level-2-managers can you please check and sign again.

FNALbuild commented:
The code-checks are being triggered in jenkins.

FNALbuild commented:
+code-checks

nitish-nayak commented:
@lgarren @mhlswang just FYI, it won't build currently. I added a tool inside the folder CVN/tools which needs a CMakeLists.txt file. Also CVN/interfaces/CMakeLists.txt might need to be modified. Please let me know if there's any other issues too and thank you for all the help!!

yangtj207 commented:
@lgarren would you help with modifying the CMakeLists.txt files? Thank you.

knoepfel commented:
@nitish-nayak, thanks for the feature branch.  We will look at the CMake files, however we notice many Python files/notebook files and an HDF5 file that probably do not belong in the PR.  Can you confirm this?

nitish-nayak commented:
@knoepfel yes, these are various scripts used for training/testing and which already existed (exists) in the dunereco version of CVN. Basically, even though the whole folder inside keras_scripts is standalone, its still needed for people to get started with CVN. Is there a better home for this than larrecodnn?

mhlswang commented:
I had a discussion with Nitish and Tingjun.  We decided that only the scripts necessary for preparing the training input files and the code used for doing inference in the LArSoft framework will be retained.  All other files such as scripts used for trainining, converting model formats, examples of previously tested models, will be removed.  A brief note will be included to describe the files remaining in the CVN folder and a link will be provided to a repository (likely the dunereco repository) that contains the files/scripts that are not included should anyone be interested in them.

FNALbuild commented:
Pull request #35 was updated. @LArSoft/level-1-managers, @LArSoft/level-2-managers can you please check and sign again.

FNALbuild commented:
The code-checks are being triggered in jenkins.

FNALbuild commented:
+code-checks

FNALbuild commented:
Pull request #35 was updated. @LArSoft/level-1-managers, @LArSoft/level-2-managers can you please check and sign again.

FNALbuild commented:
The code-checks are being triggered in jenkins.

FNALbuild commented:
+code-checks

nitish-nayak commented:
@mhlswang @lgarren I've added the README and removed the keras_scripts folder as we discussed yesterday. Thanks to Kyle as well for the CMake config! I checked that it builds now

knoepfel commented:
> @mhlswang @lgarren I've added the README and removed the keras_scripts folder as we discussed yesterday. Thanks to Kyle as well for the CMake config! I checked that it builds now

No problem @nitish-nayak.  Credit goes to @lgarren for creating the CMakeLists.txt file--I just pushed it and adjusted one of the C++ definitions.

We've made it through the CMake stuff, and now I'm going to be going through the C++.  I hope to finish my review later today.  We're getting there!

knoepfel commented on larrecodnn/CVN/func/CVNImageUtils.cxx:
```diff
@@ -0,0 +1,532 @@
+  }
+}
+
+lcvn::ViewVectorF lcvn::CVNImageUtils::ConvertViewVecToViewVecF(lcvn::ViewVector view)
+{
+
+  lcvn::ViewVectorF newVec;
+  for (size_t w = 0; w < view.size(); ++w) {
+    std::vector<float> thisWire;
+    for (size_t t = 0; t < view[w].size(); ++t) {
+      float chargeSC = static_cast<float>(view[w][t]);
+      thisWire.push_back(chargeSC);
+    }
+    newVec.push_back(thisWire);
+  }
+  return newVec;
+}
```
Here (and elsewhere), there are some inefficiencies using vectors:

- The vector passed into the function should be passed by `const` reference; otherwise there's a large expense in copying the vector into the function's stack.
- If you know what the size of the returned vector is going to be whenever the vector is created, then it's best to call `reserve(...)` on that vector to avoid unnecessary reallocations.
- Whenever you have vectors-of-vectors (which is not generally the best idea, but sometimes you're forced into it), make sure you're moving the vectors into the last container instead of copying them.

So the general pattern would look like:
```c++
std::vector<std::vector<float>> toFloats(
  std::vector<std::vector<SomethingElse>> const& view // <= by const reference
)
{
  lcvn::ViewVectorF newVec;
  newVec.reserve(view.size());  // <= reserve
  for (size_t w = 0; w < view.size(); ++w) {
    std::vector<float> thisWire;
    thisWire.reserve(view[w].size());  // <= reserve
    for (size_t t = 0; t < view[w].size(); ++t) {
      thisWire.push_back(view[w][t]);
    }
    newVec.push_back(move(thisWire));   // <= move
  }
  return newVec;
}
```

knoepfel commented on larrecodnn/CVN/func/HitType.h:
```diff
@@ -0,0 +1,27 @@
+
+namespace lcvn {
+
+  typedef enum HType {
+    kElectronHit,
+    kMuonHit,
+    kProtonHit,
+    kNeutronHit,
+    kPionHit,
+    kPiZeroHit,
+    kGammaHit,
+    kOtherPDGhit,
+    kUnknownHit,
+    kEmptyHit
+  } HitType;
```
Suggest more modern form of declaring enumerations:
```c++
enum HitType {kElectronHit, kMuonHit, ..., kEmptyHit};
```

nitish-nayak commented:
Oh, thanks to Lynn then! 

> I hope to finish my review later today.

Sounds great, let me know if you'd like me to change anything!

knoepfel commented on larrecodnn/CVN/func/HitType.cxx:
```diff
@@ -0,0 +1,3 @@
+#include "larrecodnn/CVN/func/HitType.h"
```
Suggest removing this file--it is not providing anything useful.

knoepfel commented on larrecodnn/CVN/func/InteractionType.h:
```diff
@@ -0,0 +1,165 @@
+
+namespace lcvn {
+
+  typedef enum Interaction {
```
See earlier comment about enumeration declarations.

knoepfel commented on larrecodnn/CVN/func/LinkDef.h:
```diff
@@ -0,0 +1,15 @@
+#ifdef __CINT__
```
This `LinkDef.h` file should be removed.

knoepfel commented on larrecodnn/CVN/func/PixelMap.h:
```diff
@@ -0,0 +1,96 @@
+    TH2F* ToLabTH2() const;
+    TH2F* SingleViewToTH2(const unsigned int& view) const;
+
+    unsigned int fNWire;        ///< Number of wires, length of pixel map
+    unsigned int fNTdc;         ///< Number of tdcs, width of pixel map
+    std::vector<float> fPE;     ///< Vector of PE measurements for pixels
+    std::vector<float> fPEX;    ///< Vector of X PE measurements for pixels
+    std::vector<float> fPEY;    ///< Vector of Y PE measurements for pixels
+    std::vector<float> fPEZ;    ///< Vector of Y PE measurements for pixels
+    std::vector<double> fPur;   ///< Vector of purity for pixels
+    std::vector<double> fPurX;  ///< Vector of X purity for pixels
+    std::vector<double> fPurY;  ///< Vector of Y purity for pixels
+    std::vector<double> fPurZ;  ///< Vector of Y purity for pixels
+    std::vector<HitType> fLab;  ///< Vector of Truth labels for pixels
+    std::vector<HitType> fLabX; ///< Vector of X Truth labels for pixels
+    std::vector<HitType> fLabY; ///< Vector of Y Truth labels for pixels
+    std::vector<HitType> fLabZ; ///< Vector of Y Truth labels for pixels
+    unsigned int fTotHits;      ///< Number of hits that make up the pixel map
+
+    Boundary fBound; //< Boundary of pixel map
```
How many of these can be `private`?

nitish-nayak replied:
All of them!

nitish-nayak replied:
Actually, nvm. Some of them are used elsewhere. Only `fNWire`, `fNTdc` and `fTotHits` can be made private

knoepfel commented on larrecodnn/CVN/modules/LArCVNEvaluator_module.cc:
```diff
@@ -0,0 +1,132 @@
+  {
+    produces<std::vector<lcvn::Result>>(fResultLabel);
+  }
+  //......................................................................
+  LArCVNEvaluator::~LArCVNEvaluator()
+  {
+    //======================================================================
+    // Clean up any memory allocated by your module
+    //======================================================================
+  }
+
+  //......................................................................
+  void LArCVNEvaluator::beginJob() {}
+
+  //......................................................................
+  void LArCVNEvaluator::endJob() {}
```
Suggest removing these three functions--the destructor does nothing, and because the `beginJob` and `endJob` functions are not doing anything, there's no need to provide them.

knoepfel commented on larrecodnn/CVN/interfaces/ICVNZlibMaker.h:
```diff
@@ -0,0 +1,80 @@
+
+#include "TH1.h"
+
+namespace fs = boost::filesystem;
```
This namespace alias should be removed from the header file--more specifically, from the global namespace.  Better to introduce it in the `.cxx` file that requires it.

knoepfel commented on larrecodnn/CVN/interfaces/ICVNMapper.cxx:
```diff
@@ -0,0 +1,59 @@
+    produces<std::vector<lcvn::PixelMap>>(fClusterPMLabel);
+  }
+
+  //......................................................................
+  template <class T, class U>
+  ICVNMapper<T, U>::~ICVNMapper()
+  {
+    //======================================================================
+    // Clean up any memory allocated by your module
+    //======================================================================
+  }
+
+  //......................................................................
+  template <class T, class U>
+  void ICVNMapper<T, U>::beginJob()
+  {}
+
+  //......................................................................
+  template <class T, class U>
+  void ICVNMapper<T, U>::endJob()
+  {}
```
See other comments about removing empty destructor and `beginJob`/`endJob` functions.

knoepfel commented on larrecodnn/CVN/func/InteractionType.cxx:
```diff
@@ -0,0 +1,3 @@
+#include "larrecodnn/CVN/func/InteractionType.h"
```
Suggest removing this file, which is not providing anything.

knoepfel reviewed (commented):
@nitish-nayak, thanks very much for the PR.  I have **a lot** of comments below, but please don't be discouraged by them.  I suspect that some (most?) of the code in this PR has been created over time by several people, and the comments are thus better directed toward the original author and not toward you.  For that reason, if you see any comments that you don't feel comfortable addressing, or you know you didn't write that code, then feel free to disregard.

To summarize my comments:

- It's better to remove commented-out code than to leave it in.
- Unless the destructor is intended to be virtual, don't implement an empty one.  The same holds true for _art_-related functions: don't implement empty (e.g.) `beginJob()` or `endJob()` functions.
- When using `std::vector`s, make sure you pass them to functions via reference, call `reserve(...)` when you know the size of it at creation time, and favor moving them over copying them when possible.
- Try to make as many data members `private` as possible.

I'm happy to chat with you further about any of these things.

Thanks, @nitish-nayak, for your willingness to go through this process—it can be a lot of work, but the end product is of higher quality once we go through it.

nitish-nayak commented:
@knoepfel No problem, thanks a lot for the detailed comments! I'll go through these (maybe later tonight or tomorrow) and try to incorporate all of them. It doesn't sound too complicated. And yes, a lot of this code is inherited, it's just refactored a little bit.

FNALbuild commented:
Pull request #35 was updated. @LArSoft/level-1-managers, @LArSoft/level-2-managers can you please check and sign again.

FNALbuild commented:
The code-checks are being triggered in jenkins.

FNALbuild commented:
-code-checks
Pull request failed code-formatting checks. Please ensure that `cetmodules` has been setup and execute the following command from the top-level directory of your repository:

```
format-code \
  larrecodnn/ImagePatternAlgs/Tensorflow/TF/tf_graph.cc \ 
  larrecodnn/ImagePatternAlgs/Tensorflow/TF/tf_graph.h
```
Then commit the changes and push them to your PR branch.

FNALbuild commented:
Pull request #35 was updated. @LArSoft/level-1-managers, @LArSoft/level-2-managers can you please check and sign again.

FNALbuild commented:
The code-checks are being triggered in jenkins.

FNALbuild commented:
+code-checks

nitish-nayak commented:
@knoepfel Everything should be resolved. Thanks again for the very helpful comments! Please let me know if there's any further issues

mhlswang commented:
@nitish-nayak I see that you implemented two different versions of the "run" method in tf_graph, the original and a new "runMulti" to deal with multiple inputs.  I think there is quite a bit of redundancy here that can be avoided by unifying these two methods.  I think some minor changes will need to be made in the existing models in larrecodnn (EmTrackMichelId and the 1DCNN) that make use of "run" in order to make this simplification work. I will take a look at this today.  In the meantime, can you export your CVN model in a tensorflow SavedModel format so we can also test the SavedModel implementation by Tejin Cai in tf_graph which has never been tested.  Here are some useful links that show you how to export your TF model in this format:
https://www.tensorflow.org/guide/saved_model
https://www.tensorflow.org/hub/exporting_tf2_saved_model

nitish-nayak commented:
@mhlswang Yeah I agree. Also `runMulti` returns a different type than the other `run` methods and I was hesitant to fiddle with the original code too much. 

I will look into the SavedModel format later today. Thanks again!

nitish-nayak commented:
@mhlswang Here's a tar file containing the CVN model in SavedModel format : https://file.io/4LoJAOICGne2

FNALbuild commented:
Pull request #35 was updated. @LArSoft/level-1-managers, @LArSoft/level-2-managers can you please check and sign again.

FNALbuild commented:
The code-checks are being triggered in jenkins.

FNALbuild commented:
-code-checks
Pull request failed code-formatting checks. Please ensure that `cetmodules` has been setup and execute the following command from the top-level directory of your repository:

```
format-code \
  larrecodnn/ImagePatternAlgs/Tensorflow/PointIdAlg/PointIdAlg.cxx \ 
  larrecodnn/ImagePatternAlgs/Tensorflow/TF/tf_graph.cc \ 
  larrecodnn/ImagePatternAlgs/Tensorflow/Tools/PointIdAlgTf_tool.cc \ 
  larrecodnn/ImagePatternAlgs/Tensorflow/Tools/WaveformRecogTf_tool.cc
```
Then commit the changes and push them to your PR branch.

mhlswang commented:
@nitish-nayak I just pushed my changes to your feature branch.  I unified the "run" and "runMulti" into a single method.  Please test it to see if the CVN still works after the modifications.  I also updated the SavedModel file interface. Please test if you can load the SavedModel version of the model and if it gives the same results as the ProtoBuf graph version.

FNALbuild commented:
Pull request #35 was updated. @LArSoft/level-1-managers, @LArSoft/level-2-managers can you please check and sign again.

FNALbuild commented:
The code-checks are being triggered in jenkins.

FNALbuild commented:
+code-checks

nitish-nayak commented:
@mhlswang Thanks! Will do

FNALbuild commented:
The tests are being triggered in jenkins.

FNALbuild commented:
+LArSoft tests OK on slf7 for c7:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/17432&builds= 
 +build

FNALbuild commented:
+LArSoft tests OK on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/17433&builds= 
 +build

FNALbuild commented:
-lariat failed on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:lariat/view_builds/index?offset=0&builds=lariat_ci/7328&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/17433&builds=

FNALbuild commented:
-uboone tests failed on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:uboone/view_builds/index?offset=0&builds=uboone_ci/8162&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/17433&builds=

FNALbuild commented:
-sbnd tests warning, with build warning,, with ignored warning for build, on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:sbnd/view_builds/index?offset=0&builds=sbnd_ci/10629&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/17433&builds=

FNALbuild commented:
-dune tests tests failed on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:dune/view_builds/index?offset=0&builds=dune_ci/14375&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/17433&builds=

nitish-nayak commented:
@mhlswang 
It seems one of the DUNE CI test failed with 
```
127: %MSG-s ArtException:  PostEndJob 10-Nov-2022 12:18:01 CST ModuleEndJob
128: ---- EventProcessorFailure BEGIN
129:   EventProcessor: an exception occurred during current event processing
130:   ---- ScheduleExecutionFailure BEGIN
131:     Path: ProcessingStopped.
132:     ---- OtherArt BEGIN
133:       ---- FeatureVector BEGIN
134:         Expected length: 4, provided: 3
135:         The above exception was thrown while processing module EmTrackMichelId/emtrkmichelid run: 5809 subRun: 1 event: 4503
136:       ---- FeatureVector END
137:     ---- OtherArt END
138:     Exception going through path decode
139:   ---- ScheduleExecutionFailure END
140: ---- EventProcessorFailure END
141: %MSG
```
Is this an issue?

mhlswang commented:
Hi @nitish-nayak, yes there seems to be a problem for EmTrackMichelId. I'll have to look into that.  But can you verify if it works with the CVN ?

nitish-nayak commented:
yeah, sorry. Was held up with other things. I'll try to do that today @mhlswang

mhlswang commented:
no problem @nitish-nayak , @yangtj207 just confirmed that the CVN seems to pass the test.

FNALbuild commented:
Pull request #35 was updated. @LArSoft/level-1-managers, @LArSoft/level-2-managers can you please check and sign again.

FNALbuild commented:
The code-checks are being triggered in jenkins.

FNALbuild commented:
+code-checks

FNALbuild commented:
Pull request #35 was updated. @LArSoft/level-1-managers, @LArSoft/level-2-managers can you please check and sign again.

FNALbuild commented:
The code-checks are being triggered in jenkins.

FNALbuild commented:
-code-checks
Pull request failed code-formatting checks. Please ensure that `cetmodules` has been setup and execute the following command from the top-level directory of your repository:

```
format-code \
  larrecodnn/ImagePatternAlgs/Tensorflow/TF/tf_graph.cc
```
Then commit the changes and push them to your PR branch.

yangtj207 commented:
trigger build

FNALbuild commented:
The tests are being triggered in jenkins.

FNALbuild commented:
Pull request #35 was updated. @LArSoft/level-1-managers, @LArSoft/level-2-managers can you please check and sign again.

FNALbuild commented:
The code-checks are being triggered in jenkins.

FNALbuild commented:
+code-checks

FNALbuild commented:
-argoneut tests failed on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:argoneut/view_builds/index?offset=0&builds=argoneut_ci/7630&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/17445&builds=

FNALbuild commented:
+LArSoft tests OK on slf7 for c7:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/17446&builds= 
 +build

FNALbuild commented:
-icarus tests failed on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:icarus/view_builds/index?offset=0&builds=icarus_ci/5614&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/17445&builds=

FNALbuild commented:
-sbnd tests failed on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:sbnd/view_builds/index?offset=0&builds=sbnd_ci/10633&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/17445&builds=

FNALbuild commented:
-uboone tests failed on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:uboone/view_builds/index?offset=0&builds=uboone_ci/8166&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/17445&builds=

FNALbuild commented:
-dune tests tests failed on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:dune/view_builds/index?offset=0&builds=dune_ci/14379&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/17445&builds=

lgarren commented:
Just FYI, CI tests will fail until the experiments upgrade to larsoft v09_63_00.

nitish-nayak commented:
@lgarren I went through the logs, none of the failures seem to be related to this PR

knoepfel commented:
@nitish-nayak, we think your work is probably done.  We would like to run the CI tests one more time (once experiments upgrade to the latest LArSoft release) to confirm that there are no issues.

nitish-nayak commented:
@knoepfel Awesome! thanks a lot!

yangtj207 commented:
trigger build

FNALbuild commented:
The tests are being triggered in jenkins.

FNALbuild commented:
+LArSoft tests OK on slf7 for c7:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/17498&builds= 
 +build

FNALbuild commented:
+LArSoft tests OK on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/17499&builds= 
 +build

FNALbuild commented:
-lariat failed on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:lariat/view_builds/index?offset=0&builds=lariat_ci/7342&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/17499&builds=

FNALbuild commented:
-uboone tests failed on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:uboone/view_builds/index?offset=0&builds=uboone_ci/8176&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/17499&builds=

FNALbuild commented:
-dune tests tests failed on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:dune/view_builds/index?offset=0&builds=dune_ci/14391&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/17499&builds=

FNALbuild commented:
+argoneut tests OK on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:argoneut/view_builds/index?offset=0&builds=argoneut_ci/7641&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/17499&builds=

FNALbuild commented:
-icarus tests failed on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:icarus/view_builds/index?offset=0&builds=icarus_ci/5662&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/17499&builds=

FNALbuild commented:
-sbnd tests warning on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:sbnd/view_builds/index?offset=0&builds=sbnd_ci/10654&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/17499&builds=

lgarren commented:
As expected, a number of experiments have not yet caught up with larsoft v09_63_00.  The CI tests for those experiments are failing.

yangtj207 commented:
trigger build

FNALbuild commented:
The tests are being triggered in jenkins.

FNALbuild commented:
+LArSoft tests OK on slf7 for c7:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/17533&builds= 
 +build

FNALbuild commented:
+LArSoft tests OK on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/17534&builds= 
 +build

FNALbuild commented:
-lariat failed on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:lariat/view_builds/index?offset=0&builds=lariat_ci/7346&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/17534&builds=

FNALbuild commented:
-uboone tests failed on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:uboone/view_builds/index?offset=0&builds=uboone_ci/8180&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/17534&builds=

FNALbuild commented:
+argoneut tests OK on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:argoneut/view_builds/index?offset=0&builds=argoneut_ci/7645&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/17534&builds=

FNALbuild commented:
-sbnd tests warning on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:sbnd/view_builds/index?offset=0&builds=sbnd_ci/10658&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/17534&builds=

FNALbuild commented:
-dune tests tests warning on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:dune/view_builds/index?offset=0&builds=dune_ci/14410&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/17534&builds=

FNALbuild commented:
-icarus tests warning on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:icarus/view_builds/index?offset=0&builds=icarus_ci/5683&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/17534&builds=

lgarren commented:
uboone test warnings (CI triggered by hand with v09_63_00 feature branches): 
https://dbweb8.fnal.gov:8443/LarCI/app/ns:uBooNE/view_builds/index?offset=0&builds=uboone_ci/8181&builds=

lgarren commented:
approve

FNALbuild commented:
This pull request is fully signed and it will be merged to develop and built in the next LArSoft release after it passes the integration tests.
